src/main/java/it/mulders/mcs/search/vulnerability/ComponentReportClient.java

```diff
@@ -0,0 +1,103 @@
+    this("https://ossindex.sonatype.org");
+  }
+
+  // Visible for testing
+  ComponentReportClient(final String hostname) {
+    this.hostname = hostname;
+  }
+
+  public Result<ComponentReportResponse> search(final SearchResponse.Response.Doc[] docs) {
+    return search(Stream.of(docs).map(this::toPackageUrls).toList());
+  }
+
+  public Result<ComponentReportResponse> search(final List<String> coordinates) {
+    var purls = coordinates.stream().map(purl -> "\"" + purl + "\"").collect(Collectors.joining(","));
+
+    var json = "{\"coordinates\" : [%s]}";
+
+    var builder = HttpRequest.newBuilder()
+    .version(HttpClient.Version.HTTP_1_1)
+    .POST(HttpRequest.BodyPublishers.ofString(json.formatted(purls)))
+    .header("Content-Type", "application/json");
+
+    var username = System.getProperty("ossindex.username", "");
+    var password = System.getProperty("ossindex.password", "");
+
+    if (!username.isEmpty() || !password.isEmpty()) {
+      builder.uri(URI.create(String.format("%s/api/v3/authorized/component-report", hostname)));
+      builder.header("Authorization", getBasicAuthenticationHeader(username, password));
+    } else {
+      builder.uri(URI.create(String.format("%s/api/v3/component-report", hostname)));
+      System.out.println(MISSING_OSS_INDEX_CREDENTIALS_WARNING);
+    }
+
+    try {
+      return client.send(builder.build(), new ComponentReportResponseBodyHandler()).body();
+    } catch (IOException | InterruptedException e) {
+      return new Result.Failure<>(e);
+    }
+  }
+
+  private String getBasicAuthenticationHeader(final String username, final String password) {
+    String valueToEncode = username + ":" + password;
+    return "Basic " + Base64.getEncoder().encodeToString(valueToEncode.getBytes());
+  }
+
+  public String toPackageUrls(final SearchResponse.Response.Doc doc) {
+    try {
+      return new PackageURL("maven", doc.g(), doc.a(), doc.v(), null, null).canonicalize();
+    } catch (MalformedPackageURLException e) {
+      throw new McsRuntimeException(e);
+    }
+  }
+
+  public void assignComponentReport(final ComponentReportResponse.ComponentReport componentReport,
+                                    final SearchResponse.Response.Doc[] docs) {
+    for (int i = 0; i < docs.length; i++) {
+      try {
+        PackageURL packageURL = new PackageURL(componentReport.coordinates());
+        SearchResponse.Response.Doc doc = docs[i];
+        if (matches(packageURL, doc)) {
+          doc = doc.withComponentReport(componentReport);
+          docs[i] = doc;
+          break;
+        }
+      } catch (MalformedPackageURLException e) {
+        throw new McsRuntimeException(e);
+      }
+    }
+  }
+
+  private boolean matches(final PackageURL packageURL, final SearchResponse.Response.Doc doc) {
+    return Objects.equals(packageURL.getNamespace(), doc.g()) &&
+        Objects.equals(packageURL.getName(), doc.a()) &&
+        Objects.equals(packageURL.getVersion(), doc.v());
```

Comment:
When the user performs a search that does not return one exact coordinate, `doc.v()` is empty. In that case, I think the best we could do is use `doc.latestVersion()`.

Reply:
> When the user performs a search that does not return one exact coordinate, `doc.v()` is empty. In that case, I think the best we could do is use `doc.latestVersion()`.

608650f3